core/src/main/java/io/grpc/internal/ClientCallImpl.java

```diff
@@ -366,7 +391,7 @@ public void cancel(@Nullable String message, @Nullable Throwable cause) {
 
   @Override
   public void halfClose() {
-    Preconditions.checkState(stream != null, "Not started");
```

Comment:
Leave this as-is. It could fail if they didn't write any message before half-close, and "not started" would not be an appropriate error message. It could be a new check, although it'd probably be better to not throw an exception and instead fail the RPC with a status.